Initial TypeScript setup

fgblomqvist opened the pull request:
This PR demonstrates what a very basic/initial TS setup would look like. The purpose is for it to act as a base for discussions until we have reached a satisfied solution. It only sets up TS in `beta` and `button` for now, to keep things minimal.

Changes:
* Add a basic tsconfig.json which is setup to be very lenient and allow us to incrementally add types as-we-go
  * The goal should be to stricten the rules over time to catch more bugs and make the code even easier to follow
* The flow will now be:
  1. Compile the src to JS and copy any non-JS files, output everything to `build`
  2. Run `jalla` on files in `build` to serve them or bundle/transpile them further
* `jalla` doesn't seem to like running from a subfolder, so we have to cd into `build` before running it
  * This also means we have to copy the package.json into the build folder, since `jalla` reads its config from there and you cannot specify a custom path to it
* With this configuration, we can now write ESM code that will then get transpiled into CJS (for now) that jalla can read
  * The goal should be to configure jalla/browserify to read ESM and also output ESM (it's 2022, it's supported by all browsers)
* Even if TS spits out errors, the code will still build
  * This is good as a first step since it means TS won't be a blocker, however, in the long-term it should be switched to simply not build if there are errors

To convert a file to TS, you simply change its extension from `.js` to `.ts`. Doing that will most likely reveal some basic errors. I have done it to two files here to demo what it looks like, and it did indeed find some code that needed to be fixed. After a file is in TS, you can add types inside of it and also use ESM import syntax. If an imported module has types, they will be used. If it doesn't, then it will simply be untyped.

On a side note, one of the files I converted I don't understand the purpose of: the load-timeout one. What is it supposed to achieve? It looks like it's supposed to return a promise that resolves to a timeout....once the timeout has resolved? But what's the point of that? The way it's used is by awaiting the return and _then_ cancelling the timeout. Which is not possible, and also doesn't make any sense. Maybe I'm missing something, but I changed it to what I'm guessing it should be, but its usages will have to be updated :slightly_smiling_face:. Curious to hear if anyone knows more about this!

fgblomqvist commented:
As a first comment, we could do the file move (into `src`) as a separate PR before this to make this one clearer. If someone disagrees about the move I'm all ears!

peterklingelhofer commented:
Great work on this so far @fgblomqvist. I've found that [MakeTypes](https://jvilk.com/MakeTypes) can speed up the interface creation process a good bit when converting projects (probably pretty common knowledge but figured I'd post this for those who hadn't seen it yet).

fgblomqvist commented:
Thanks for the tip @peterklingelhofer, I actually didn't know about that tool, but looks like it'll indeed be handy 🙂. Something we might want to look into as well is if we can generate typed API clients, but that could be a later thing.

fgblomqvist commented:
I think it makes sense to consider/merge https://github.com/resonatecoop/stream/pull/163 first, before adding TS. Just to deal with one thing at a time.

fgblomqvist commented:
I want to add TS to one of the other packages as well, such as `button`, to demonstrate what that would look like, but I'm gonna hold off until a decision on #163 is made since that impacts things.

fgblomqvist commented:
Need to do a few more things to this before it's ready for another round of review. I rebased on latest develop with npm workspaces, and changed the TS setup to be repo-wide so that we can do "tsc --watch" on any and all ts packages at the same time (e.g. just running a single tsc --watch rather than one per package). Just need to make sure all the commands jive together and work out any kinks, shouldn't be too much more to do.

fgblomqvist commented:
This PR has grown quite a bit while I've tried to iron out the last issues (not sure why the tests are failing in the pipeline, they're actually passing locally now).

I think it would be better to split this into a few smaller ones:
* Move files in beta into a `src` folder (and update jalla config etc. accordingly)
* Switch from `standard` to `eslint` with https://github.com/standard/eslint-config-standard
  * Makes it more seamless to switch to the TS config in this MR
* Update dependency-check
  * There's a v5 pre-release that can handle TS files natively
* Potentially remove `sheetify` (if it's not needed/being used)

All of these are quite straight-forward on their own. I might proceed with a few of them in the coming days.

fgblomqvist commented:
Rebased this on top of #174 for now.

fgblomqvist commented on .eslintrc.js:
```diff
@@ -1,11 +1,35 @@
+  overrides: [{
+    files: ['*.ts', '*.tsx'],
+    rules: {
+      '@typescript-eslint/restrict-template-expressions': ['error', {
+        allowNumber: true,
+        allowBoolean: false,
+        allowAny: true,
+        allowNullish: true,
+        allowRegExp: false
+      }],
+      '@typescript-eslint/strict-boolean-expressions': ['error', {
+        allowString: true,
+        allowNumber: true,
+        allowNullableObject: true,
+        allowNullableBoolean: true,
+        allowNullableString: true,
+        allowNullableNumber: true,
+        allowAny: true
+      }]
+    }
+  }]
```
I made these a little looser since otherwise we'll get lots of warnings for all the not yet typed code.

fgblomqvist commented on packages/button/index.ts:
```diff
@@ -69,4 +87,4 @@ function Button (props = {}) {
 }
 
-module.exports = Button
+export = Button
```
This is how we'll have to export the converted components until everything has switched to ESM. Once that happens, we can use normal ESM exports. However, this also means that we are limited to a single export per file until then. Just something to keep in mind.

fgblomqvist commented on package.json:
```diff
@@ -36,8 +39,10 @@
+    "dependency-check": "^5.0.0-4",
     "eslint": "^7.32.0",
-    "eslint-config-standard": "^16.0.3",
-    "npm-run-all": "^4.1.5"
+    "eslint-config-standard-with-typescript": "^21.0.1",
+    "npm-run-all": "^4.1.5",
+    "typescript": "^4.5.5"
```
There will be a warning when running eslint on the TS code saying that we're using too new of a TS version. This is because TS 4.5 isn't officially supported until ESLint 8. However, `standard` doesn't support ESLint 8 yet so we'll have to wait on them to upgrade (should happen in the next months). Regardless, the warning is harmless until then.

fgblomqvist commented on beta/src/types.ts:
```diff
@@ -0,0 +1,88 @@
+// Utility type for defining that you want a class, rather than an instance of a class
+type Class<T> = new (...args: any[]) => T
+
+export interface AppState extends Pick<IState, KnownKeys<IState>> {
```
I did my best in speccing this out. It's gonna have to be a WIP that will improve as more and more code gets brought into TS-land. Theoretically, it'll correct itself the more types we add (or bugs/irrational behavior will be exposed).

fgblomqvist commented on beta/src/stores/tracks.types.ts:
```diff
@@ -0,0 +1,20 @@
+export interface Track {
```
Had to put these in a separate file since we can't export multiple things from the component files for now (see other comment).

fgblomqvist commented:
@auggod @jackhajb This should be ready for review now. I'm satisfied myself at least. Open to discuss or explain further any of the parts.

auggod commented:
I'll test locally. I'm eager to merge too : )
